sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala

```diff
@@ -0,0 +1,226 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.orc
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.hadoop.io._
+import org.apache.orc.mapred.{OrcList, OrcMap, OrcStruct, OrcTimestamp}
+import org.apache.orc.storage.serde2.io.{DateWritable, HiveDecimalWritable}
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.SpecificInternalRow
+import org.apache.spark.sql.catalyst.util._
+import org.apache.spark.sql.execution.datasources.orc.OrcUtils.withNullSafe
+import org.apache.spark.sql.types._
+import org.apache.spark.unsafe.types.UTF8String
+
+private[orc] class OrcDeserializer(
+    dataSchema: StructType,
+    requiredSchema: StructType,
+    missingColumnNames: Seq[String]) {
+
+  private[this] val mutableRow = new SpecificInternalRow(requiredSchema.map(_.dataType))
+
+  private[this] val length = requiredSchema.length
+
+  private[this] val unwrappers = requiredSchema.map { f =>
+    if (missingColumnNames.contains(f.name)) {
+      (value: Any, row: InternalRow, ordinal: Int) => row.setNullAt(ordinal)
+    } else {
+      unwrapperFor(f.dataType)
+    }
+  }.toArray
+
+  def deserialize(orcStruct: OrcStruct): InternalRow = {
+    val names = orcStruct.getSchema.getFieldNames
+    val fieldRefs = requiredSchema.map { f =>
+      val name = f.name
+      if (missingColumnNames.contains(name)) {
+        null
+      } else {
+        if (names.contains(name)) {
+          orcStruct.getFieldValue(name)
+        } else {
+          orcStruct.getFieldValue("_col" + dataSchema.fieldIndex(name))
```

Comment:
actually, do we really need `missingCols`? If the name is missing, `orcStruct.getFieldValue` will just return null, isn't it?

Reply:
`OrcStruct` will raise Exception here.

Reply:
Then how can we guarantee this line won't throw exception?

Reply:
We handled them in a missing columns in line 55.